apps/prairielearn/src/lib/cache.ts

```diff
@@ -1,138 +1,132 @@
+    }
+
+    case 'redis': {
       // This returns a promise, but we don't want to wait for this data
       // to reach the cache before continuing, and we don't *really*
-      // care if it errors
-      if (maxAgeMS) {
-        // we don't log the error because it contains the cached value, which can be huge and which fills up the logs
-        cache
-          .set(key, value, maxAgeMS)
-          .catch((_err) => logger.error('Cache set error', { key, maxAgeMS }));
-      } else {
-        // we don't log the error because it contains the cached value, which can be huge and which fills up the logs
-        cache.set(key, value).catch((_err) => logger.error('Cache set error', { key }));
-      }
+      // care if it errors.
+      //
+      // We don't log the error because it contains the cached value,
+      // which can be huge and which fills up the logs.
+      client
+        .set(key, JSON.stringify(value), { PX: maxAgeMS ?? undefined })
+        .catch((_err) => logger.error('Cache set error', { key, maxAgeMS }));
       break;
-    default:
-      throw new Error(`Unknown cache type "${cacheType}"`);
+    }
   }
 }
 
 /**
- * Calls the callback with the value if it exists in the cache, or null otherwise.
+ * Returns the value for the corresponding key if it exists in the cache; null otherwise.
  */
-export async function getAsync(key: string) {
+export async function get(key: string): Promise<any> {
   if (!cacheEnabled) return null;
 
   switch (cacheType) {
-    case 'memory':
-      return cache.get(key) ?? null;
-    case 'redis':
-      return cache.get(key);
-    default:
-      throw new Error(`Unknown cache type "${cacheType}"`);
+    case 'memory': {
+      const value = cache.get(key);
+      if (typeof value === 'string') {
+        return JSON.parse(value);
+      }
+      return null;
+    }
+
+    case 'redis': {
+      const value = await client.get(key);
+      if (typeof value === 'string') {
+        return JSON.parse(value);
+      }
+      return null;
+    }
+
+    default: {
+      return null;
+    }
   }
 }
 
-export const get = util.callbackify(getAsync);
-
 /**
- * Clears all entries from the cache. Mostly used to avoid leaking memory
- * during testing.
+ * Clear all entries from the cache.
  */
-export function reset(callback: (err?: Error | null) => void) {
-  if (!cacheEnabled) return callback(null);
+export async function reset() {
+  if (!cacheEnabled) return;
 
   switch (cacheType) {
-    case 'memory':
+    case 'memory': {
       cache.clear();
-      callback(null);
       break;
-    case 'redis':
-      cache
-        .reset()
-        .then(() => callback(null))
-        .catch(callback);
+    }
+
+    case 'redis': {
+      let cursor = 0;
+      do {
+        const reply = await client.scan(cursor, { MATCH: '*', COUNT: 1000 });
+        cursor = reply.cursor;
+        console.log('deleting keys', reply.keys);
+        await client.del(reply.keys);
+      } while (cursor !== 0);
```

Comment:
I stole this from Rails:

https://github.com/rails/rails/blob/e88857bbb9d4e1dd64555c34541301870de4a45b/activesupport/lib/active_support/cache/redis_cache_store.rb#L210-L229

Reply:
Cute!